rustls/examples/internal/test_ca.rs

```diff
@@ -0,0 +1,365 @@
+                        cert: issuer,
+                        key_pair: issuer_key,
+                    } = certified_keys
+                        .get(&(Role::TrustAnchor, alg))
+                        .unwrap();
+                    role.params(alg)
+                        .signed_by(&key_pair, issuer, issuer_key)
+                        .unwrap()
+                }
+                Role::EndEntity | Role::Client => {
+                    let CertifiedKey {
+                        cert: issuer,
+                        key_pair: issuer_key,
+                    } = certified_keys
+                        .get(&(Role::Intermediate, alg))
+                        .unwrap();
+                    role.params(alg)
+                        .signed_by(&key_pair, issuer, issuer_key)
+                        .unwrap()
+                }
+            };
+            // Serialize the issued certificate to a PEM encoded file.
+            let mut cert_file = File::create(role.cert_pem_file_path(alg)).unwrap();
+            cert_file
+                .write_all(cert.pem().as_bytes())
+                .unwrap();
+            // And to a DER encoded file.
+            let mut cert_file = File::create(role.cert_der_file_path(alg)).unwrap();
+            cert_file.write_all(cert.der()).unwrap();
+
+            certified_keys.insert((role, alg), CertifiedKey { cert, key_pair });
+        }
+    }
+
+    // Write the PEM chain and full chain files for the end entity and client certs.
+    // Chain files include the intermediate and root certs, while full chain files include
+    // the end entity or client cert as well.
+    for role in [Role::EndEntity, Role::Client] {
+        for alg in sig_algs {
+            let CertifiedKey { cert: root, .. } = certified_keys
+                .get(&(Role::TrustAnchor, alg))
+                .unwrap();
+            let CertifiedKey {
+                cert: intermediate, ..
+            } = certified_keys
+                .get(&(Role::Intermediate, alg))
+                .unwrap();
+            let CertifiedKey {
+                cert: end_entity, ..
+            } = certified_keys
+                .get(&(role, alg))
+                .unwrap();
+            fn write_chain_file(
+                role: Role,
+                alg: &'static SignatureAlgorithm,
+                name: &str,
+                chain_certs: &[&Certificate],
+            ) {
+                let mut chain_file =
+                    File::create(output_directory(alg).join(format!("{}.{}", role.label(), name)))
+                        .unwrap();
+                for cert in chain_certs {
+                    chain_file
+                        .write_all(cert.pem().as_bytes())
+                        .unwrap();
+                }
+            }
+
+            write_chain_file(role, alg, "chain", &[intermediate, root]);
+            write_chain_file(role, alg, "fullchain", &[end_entity, intermediate, root]);
```

Comment:
Instead of a function, consider a loop over `let files = [("chain", &[intermediate, root]), ("fullchain", &[end_entity, intermediate, root])];"?

Reply:
Done.